### PR TITLE
Add load_account_with() API to allow a callback to control whether to insert the loaded account into read_cache.

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

https://github.com/anza-xyz/agave/pull/1157

Split the accounts-db related changes from #1157. The accounts-db changes don't depend on #1157. The new API can be used for other cases than #1157. It also makes the code review easier.  

#### Summary of Changes

- Add a new API in accounts-db - `load_with()`, which takes a callback to indicate whether to insert the loaded account into read_cache.
- Add a test for the new API with read_cache.

Fixes #

### CODE REVIEW

File: accounts-db/src/accounts_db.rs
Comment (Lichtso) [inline accounts-db/src/accounts_db.rs:5468]: Looks good. This should work (for the new index structure of the program cache).

File: accounts-db/src/accounts_db.rs
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: Looks like there's a lot of duplication with `do_load_with_populate_read_cache()`. Can they be refactored at all?

File: accounts-db/src/accounts_db.rs
Comment (Lichtso) [reply accounts-db/src/accounts_db.rs:5468]: [refactor/last_written_slot_in_program_loading](https://github.com/Lichtso/solana/tree/refactor/last_written_slot_in_program_loading) is where it will be used. Once this PR merges I can rebase that branch and once that is merged @HaoranYi could rebase #1157. That would be one way to make the two change sets compatible.

File: accounts-db/src/accounts_db.rs
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: The pattern for these two functions is similar. But the function signatures are very different. Because of the function signature difference, almost every code if branches inside the function body is a little different. this makes sharing the code difficult. Therefore, I think it is probaby okay to keep them separate. WDYT?

Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs]: Gotcha. I'll look again. If I think of anything useful, I'll let you know.

Comment (Lichtso) [reply accounts-db/src/accounts_db.rs]: We could potentially make `do_load_with_populate_read_cache(key)` an alias for `load_account_with(key, |_| true)`.

File: accounts-db/src/accounts_db.rs
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: Because this short-circuits, it means that if the account is in the write cache, we will *not* call `callback`. This is surprising to me, but maybe that's the intention for this function? If yes, I think we should document it better. Also, wdyt about changing the name of the parameter to something like `should_put_in_read_cache_fn`?

The documentation on the function itself can say something like:
```rust
/// Load account with `pubkey` and maybe put into read cache.
///
/// If the account is not already cached, invoke `should_put_in_read_cache_fn`.
/// The caller can inspect the account and indicate if it should be put into the read cache or not.
///
/// Return the account and the slot when the account was last stored.
/// Return None for ZeroLamport accounts.
```

File: accounts-db/src/read_only_accounts_cache.rs
Comment (brooksprumo) [inline accounts-db/src/read_only_accounts_cache.rs]: Can this be decorated with DCOU? And then in that case, I think `pub` is fine too.

```suggestion
        /// reset the read only accounts cache
        #[cfg(feature = "dev-context-only-utils")]
        pub fn reset_for_tests(&self) {
```

File: accounts-db/src/accounts_db.rs
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:13671]: I think there should be a scenario to test when the account is in the write cache. In that case, we should *not* add it to the read cache, regardless of what the callback function says.

File: accounts-db/src/accounts_db.rs
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: yes. done.

File: accounts-db/src/read_only_accounts_cache.rs
Comment (HaoranYi) [reply accounts-db/src/read_only_accounts_cache.rs]: yes. done.

File: accounts-db/src/accounts_db.rs
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs:13671]: yes. added.

File: accounts-db/src/accounts_db.rs
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: nit: typo
```suggestion
        // The account shouldn't be added to read_only_cache because it is in write_cache.
```

IMO this doesn't need to block merging; it can be addressed in a different PR (if that's preferred). However, if there are other changes that happen, please pull in this fix.

**Review by brooksprumo — DISMISSED** [review]
:shipit:

**Comment by Lichtso** [comment]
Anything else holding this PR back? Otherwise I would say fix the typo, we reapprove and merge.

**Comment by HaoranYi** [comment]
Fix a typo.

**Review by brooksprumo — APPROVED** [review]

**Pull request merged**
